bin/s3stalker_daemon.py

```diff
@@ -0,0 +1,176 @@
+
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""S3stalker daemon/runner.
+
+This is a daemon supposed to stay up and running "forever". It will continously
+fetch fresh filenames from an S3 object store and publish the urls of those new
+filenames. It is a daemon version of the s3stalker.py script which needs to be
+run as a cronjob.
+
+"""
+
+import argparse
+import logging
+import logging.config
+from datetime import datetime, timedelta, timezone
+import time
+import yaml
+import sys
+
+from pytroll_collectors.helper_functions import read_yaml
+from pytroll_collectors.s3stalker import create_messages_for_recent_files
+from pytroll_collectors.s3stalker import get_last_fetch
+
+import signal
+from threading import Thread
+from posttroll.publisher import create_publisher_from_dict_config
+
+logger = logging.getLogger(__name__)
+
+
+class S3StalkerRunner(Thread):
+    """Runner for stalking for new files in an S3 object store."""
+
+    def __init__(self, bucket, config, startup_timedelta_seconds):
+        """Initialize the S3Stalker runner class."""
+        super().__init__()
+
+        self.bucket = bucket
+        self.config = config
+        self.startup_timedelta_seconds = startup_timedelta_seconds
+        self.time_back = self.config['timedelta']
+        self.publisher = None
+        self.loop = False
+        self._setup_and_start_communication()
+
+    def _setup_and_start_communication(self):
+        """Set up the Posttroll communication and start the publisher."""
+        logger.debug("Starting up... ")
+        self.publisher = create_publisher_from_dict_config(self.config['publisher'])
+        self.publisher.start()
+        self.loop = True
+        signal.signal(signal.SIGTERM, self.signal_shutdown)
+
+    def signal_shutdown(self, *args, **kwargs):
+        """Shutdown the S3 Stalker daemon/runner."""
+        self.close()
+
+    def run(self):
+        """Start the s3-stalker daemon/runner in a thread."""
+        first_run = True
+        config = self.config
+        last_fetch_time = None
+        while self.loop:
+            if first_run:
+                logger.info('Create messages with urls for most recent files only')
+                logger.info('On start up we consider files with age up to %d seconds from now',
+                            self.startup_timedelta_seconds)
+                config['timedelta'] = {'seconds': self.startup_timedelta_seconds}
+                first_run = False
+            else:
+                seconds_back = self._get_seconds_back_to_search(last_fetch_time)
+                config['timedelta'] = {'seconds': seconds_back}
```

Comment:
This could be refactored to a separate method, and used something like this:
```python
config['timedelta'] = self._get_timedelta(last_fetch_time, is_first_run=first_run)
first_run = False
```

Reply:
Done